Update device selection explainer (#824)
webmachinelearning/webnn: device-selection-explainer.md

Implementations, browsers, and the underlying OS may want to control the allocation of compute devices for various use cases and system conditions.

> The question is in what use cases who and to what extent should control the execution context.

Reviewer: ```suggestion
The question is who should be able to, and to what extent, control the execution context state and capabilities.
```

- Discuss option #3 from [Considered alternatives](#considered-alternatives).

## Next Phase Device Selection Solution

In [Remove MLDeviceType #809](https://github.com/webmachinelearning/webnn/pull/809)  this [comment](https://github.com/webmachinelearning/webnn/pull/809#discussion_r1936856070) raised a new use case:

> about the likely need for a caller to know whether a particular device is supported or not, because an app may want to (if say GPU is not supported) use a different more performant fallback than for WebNN to silently fall back to CPU. For example, if GPU was unavailable (even though you preferred high performance), then it might be faster to execute the model with WebGPU shaders than WebNN CPU, or it might be okay to use CPU, but the app could load a different model that's more CPU-friendly, if it knew that was the case.

> That sparked a discussion in [Query mechanism for supported devices #815

Reviewer: This use case should be added here: https://github.com/webmachinelearning/webnn/issues/815#issuecomment-2658627753

When we go deeper to the solutions space a bit later (hoping we're able to extract some more use cases!), querySupport could go into considered alternatives along with a description of its known issues.

> That sparked a discussion in [Query mechanism for supported devices #815

Author: OK, I will add the other use cases and reformulate the specific idea as an example to a starting point that could be explored next.

That sparked a discussion in [Query mechanism for supported devices #815
](https://github.com/webmachinelearning/webnn/issues/815) about possible solutions, for instance [this shape](https://github.com/webmachinelearning/webnn/issues/815#issuecomment-2657101952) emerged as a possible starting point for further exploration, drafting a generic mechanism for capability introspection and examples of possible parameters and outcomes.

```js
const support = await context.querySupport({
  dataTypes: ['float16', 'int8'],
  maximumRank: 6,
  operators: ['lstm', 'hardSwish'],
});
console.log(support); // "optimized" or "fallback"
```

The next phase in developing device selection is therefore to explore this proposal and eventually others.

> Other use cases were raised as well, in this comment for realtime video processing:

Reviewer: ```suggestion
Other use cases were raised as well, in [this comment](https://github.com/webmachinelearning/webnn/issues/815#issuecomment-2658627753) for realtime video processing:
```
